pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -574,33 +666,51 @@ func getVolumesFromPodDir(podDir string) ([]podVolume, error) {
 		}
 		podName := podsDirInfo[i].Name()
 		podDir := path.Join(podDir, podName)
-		volumesDir := path.Join(podDir, config.DefaultKubeletVolumesDirName)
-		volumesDirInfo, err := ioutil.ReadDir(volumesDir)
-		if err != nil {
-			glog.Errorf("Could not read volume directory %q: %v", volumesDir, err)
-			continue
-		}
-		for _, volumeDir := range volumesDirInfo {
-			pluginName := volumeDir.Name()
-			volumePluginPath := path.Join(volumesDir, pluginName)
 
-			volumePluginDirs, err := utilfile.ReadDirNoStat(volumePluginPath)
-			if err != nil {
-				glog.Errorf("Could not read volume plugin directory %q: %v", volumePluginPath, err)
+		// TODO: remove feature gate check after no longer needed
+		volumesDirs := []string{path.Join(podDir, config.DefaultKubeletVolumesDirName)}
+		if utilfeature.DefaultFeatureGate.Enabled(features.BlockVolume) {
+			// Find both filesystem volume and block volume information
+			// ex. filesystem volume: /pods/{podUid}/volume/{escapeQualifiedPluginName}/{volumeName}
+			//     block volume:      /pods/{podUid}/volumeDevices/{escapeQualifiedPluginName}/{volumeName}
+			volumesDirs = append(volumesDirs, path.Join(podDir, config.DefaultKubeletVolumeDevicesDirName))
+		}
+		for _, volumesDir := range volumesDirs {
+			var volumesDirInfo []os.FileInfo
+			if volumesDirInfo, err = ioutil.ReadDir(volumesDir); err != nil {
+				// Just skip the loop becuase given volumesDir doesn't exist depending on volumeMode
 				continue
 			}
-
-			unescapePluginName := strings.UnescapeQualifiedNameForDisk(pluginName)
-			for _, volumeName := range volumePluginDirs {
-				mountPath := path.Join(volumePluginPath, volumeName)
-				volumes = append(volumes, podVolume{
-					podName:        volumetypes.UniquePodName(podName),
-					volumeSpecName: volumeName,
-					mountPath:      mountPath,
-					pluginName:     unescapePluginName,
-				})
+			for _, volumeDir := range volumesDirInfo {
+				pluginName := volumeDir.Name()
+				volumePluginPath := path.Join(volumesDir, pluginName)
+				volumePluginDirs, err := utilfile.ReadDirNoStat(volumePluginPath)
+				if err != nil {
+					glog.Errorf("Could not read volume plugin directory %q: %v", volumePluginPath, err)
+					continue
+				}
+				// TODO: remove feature gate check after no longer needed
+				var volumeMode v1.PersistentVolumeMode
+				if utilfeature.DefaultFeatureGate.Enabled(features.BlockVolume) {
+					if volumesDir == path.Join(podDir, config.DefaultKubeletVolumeDevicesDirName) {
```

Comment:
Can the mode be specified when defining the volumeDirs above?  I think it should be ok to specify v1.PersistentVolumeFilesystem outside of the feature gate in this particular scenario because it will be ignored when we're reading it.

Reply:
I think we did this in previous. ack